### PR TITLE
External id removal

### PR DESCRIPTION
**Pull request by ZakarFin**
Removes references to oskari_maplayer.external_id and usage of string typed layer ids when referencing layers in oskari_maplayer.

Note! There's extensive refactoring of InsertFeature, SaveFeature and DeleteFeature routes under control-base as they were changed a bit for the string based layer id handling.

### CODE REVIEW

File: control-admin/src/main/java/fi/nls/oskari/control/admin/SearchWFSChannelActionHandler.java [hunk 1]
```diff
@@ -60,9 +60,9 @@ public void handleGet(ActionParameters params)
         try {
             for(WFSSearchChannelsConfiguration channel : channelService.findChannels()) {
                 JSONObject channelJSON = channel.getAsJSONObject();
-                List<String> layerIds = new ArrayList<String>();
-                layerIds.add(String.valueOf(channel.getWFSLayerId()));
-                JSONObject userLayers = OskariLayerWorker.getListOfMapLayersById(layerIds, params.getUser(), params.getLocale().getLanguage(), params.getHttpParam(PARAM_SRS));
+                List<Integer> layerIds = new ArrayList<>();
+                layerIds.add(channel.getWFSLayerId());
```
Comment (jampukka) [inline control-admin/src/main/java/fi/nls/oskari/control/admin/SearchWFSChannelActionHandler.java]: or `Collections.singletonList(channel.getWFSLayerId());`

File: control-base/src/main/java/fi/nls/oskari/control/data/GetPrintHandler.java [hunk 2]
```diff
@@ -211,7 +211,11 @@ private void setFormat(String formatStr, PrintRequest req)
 
         List<PrintLayer> printLayers = new ArrayList<>(requestedLayers.length);
         for (LayerProperties requestedLayer : requestedLayers) {
-            OskariLayer oskariLayer = permissionHelper.getLayer(requestedLayer.getId(), user);
+            int id = ConversionHelper.getInt(requestedLayer.getId(), -1);
+            if (id == -1) {
+                continue;
```
Comment (jampukka) [inline control-base/src/main/java/fi/nls/oskari/control/data/GetPrintHandler.java]: myplaces_, userlayer_, analysis_ layers won't be sent to the PrintService after this. However, as PrintService doesn't currently support those types of layers this isn't a big deal.

File: control-base/src/main/java/fi/nls/oskari/control/feature/DeleteFeatureHandler.java [hunk 3]
```diff
@@ -91,6 +93,14 @@ else if (responseString.indexOf("<wfs:totalDeleted>1</wfs:totalDeleted>") > -1)
 			throw new ActionException("Could not update feature");
 		}
 	}
+
+	private int getLayerId(String layerId) throws ActionParamsException {
```
Comment (jampukka) [inline control-base/src/main/java/fi/nls/oskari/control/feature/DeleteFeatureHandler.java]: This same method seems to be implemented in several different locations with varying error messages. Maybe we should replace them with a single static function?

File: control-base/src/main/java/fi/nls/oskari/control/data/GetPrintHandler.java (hunk 2 above)
Comment (ZakarFin) [reply control-base/src/main/java/fi/nls/oskari/control/data/GetPrintHandler.java]: True, having the support for those layers is another matter :) Previously the method just failed as it didn't find the referenced layers in oskari_maplayer.

Comment (jampukka) [reply control-base/src/main/java/fi/nls/oskari/control/data/GetPrintHandler.java]: Yea

File: control-base/src/main/java/fi/nls/oskari/control/feature/DeleteFeatureHandler.java (hunk 3 above)
Comment (ZakarFin) [reply control-base/src/main/java/fi/nls/oskari/control/feature/DeleteFeatureHandler.java]: Got a little side-tracked and rewrote the these handlers to some clarity. They should be further updated to use the wfs-t code found in service-myplaces.

File: control-base/src/main/java/fi/nls/oskari/control/feature/AbstractFeatureHandler.java [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        writer.append("<gml:MultiLineString xmlns:gml=\"http://www.opengis.net/gml\" srsName=\"");
+        writer.append(srsName);
+        writer.append("\">");
+        for (int lineIndex = 0; lineIndex < data.length(); lineIndex++) {
+            writer.append("<gml:lineStringMember><gml:LineString><gml:coordinates decimal=\".\" cs=\",\" ts=\" \">");
+            JSONArray lineCoordinates = data.getJSONArray(lineIndex);
+            for (int coordinateIndex = 0; coordinateIndex < lineCoordinates.length(); coordinateIndex++) {
+                JSONObject coord = lineCoordinates.getJSONObject(coordinateIndex);
+                writer.append(coord.getString("x") + "," + coord.getString("y"));
+                if (coordinateIndex < (lineCoordinates.length() - 1)) {
```
Comment (jampukka) [inline control-base/src/main/java/fi/nls/oskari/control/feature/AbstractFeatureHandler.java]: Or move to the start of the loop and check `coordinateIndex > 0`. But don't bother changing it, there's too much to fix...

File: control-base/src/main/java/fi/nls/oskari/control/feature/DeleteFeatureHandler.java [hunk 5]
```diff
@@ -1,105 +1,63 @@
...
+            throw new ActionParamsException("Error getting featureId from payload");
+        }
+    }
 
-		layerService = new OskariLayerServiceIbatisImpl();
-		permissionsService = new PermissionsServiceIbatisImpl();
-		layerConfigurationService = new WFSLayerConfigurationServiceIbatisImpl();
-		String featureData = params.getHttpParam("featureData");
+    protected String createPayload(String layerName, String featureId) {
+        // TODO: rewrite
```
Comment (jampukka) [inline control-base/src/main/java/fi/nls/oskari/control/feature/DeleteFeatureHandler.java]: Yea...

File: control-base/src/main/java/fi/nls/oskari/control/feature/DeleteFeatureHandler.java [hunk 6]
```diff
@@ -1,105 +1,63 @@
...
+        StringBuilder requestData = new StringBuilder(
+                "<wfs:Transaction service='WFS' version='1.1.0'" +
+                        " xmlns:ogc='http://www.opengis.net/ogc'" +
+                        " xmlns:wfs='http://www.opengis.net/wfs'>" +
+                        "<wfs:Delete typeName='" + layerName + "'>" +
+                            "<ogc:Filter><ogc:FeatureId fid='" + featureId + "'/></ogc:Filter>" +
+                        "</wfs:Delete></wfs:Transaction>");
+        return requestData.toString();
```
Comment (jampukka) [inline control-base/src/main/java/fi/nls/oskari/control/feature/DeleteFeatureHandler.java]: Maybe `return requestData.toString().replace('\'', '"');`

File: control-base/src/main/java/fi/nls/oskari/control/feature/InsertFeatureHandler.java [hunk 7]
```diff
@@ -47,100 +26,37 @@
...
-    private JSONObject getJSONPayload(ActionParameters params)
-            throws ActionException {
+        JSONObject jsonPayload = params.getHttpParamAsJSON("featureData");
+        OskariLayer layer = getLayer(jsonPayload.optString("layerId"));
 
-        String featureData = params.getHttpParam("featureData");
-        JSONObject jsonObject = JSONHelper.createJSONObject(featureData);
-        if (jsonObject == null) {
-            throw new ActionParamsException("Couldn't parse featureData JSON from request");
+        if (!canEdit(layer, params.getUser())) {
```
Comment (jampukka) [inline control-base/src/main/java/fi/nls/oskari/control/feature/InsertFeatureHandler.java]: DeleteFeatureHandler checks `params.requireLoggedInUser();` prior to calling this, I like it.

File: control-base/src/main/java/fi/nls/oskari/control/feature/InsertFeatureHandler.java [hunk 8]
```diff
@@ -47,100 +26,37 @@
...
-        httpClientBuilder.setDefaultCredentialsProvider(credsProvider);
-        return httpClientBuilder.build();
+        flushLayerTilesCache(layer.getId());
+        ResponseHelper.writeResponse(params, JSONHelper.createJSONObject("fid", updatedFeatureId));
     }
 
-    private String createWFSTMessage(JSONObject jsonPayload, OskariLayer layer)
+    private String createWFSTMessage(JSONObject jsonPayload, OskariLayer layer, WFSLayerConfiguration wfsMetadata)
             throws ActionException {
         final String srsName = JSONHelper.getStringFromJSON(jsonPayload, "srsName", "http://www.opengis.net/gml/srs/epsg.xml#3067");
```
Comment (jampukka) [inline control-base/src/main/java/fi/nls/oskari/control/feature/InsertFeatureHandler.java]: Love the hard-coded-default-to-3067

**Comment by jampukka** [comment]
Changes look good. There's a slight chance this breaks something, but the benefits outweigh the risks.

**Pull request merged**